plotly/graph_objs/graph_objs.py

```diff
@@ -297,67 +348,175 @@ def force_clean(self, caller=True):
+
+    def __setitem__(self, key, value, _raise=True):
+
+        if not isinstance(key, six.string_types):
+            if _raise:
+                raise TypeError('Key must be string, not {}'.format(type(key)))
+            return
+
+        if key.endswith('src') and key in self._attributes:
+            value = graph_objs_tools.assign_id_to_src(key, value)
+            return super(PlotlyDict, self).__setitem__(key, value)
+
+        subplot_key = self._get_subplot_key(key)
+        if subplot_key is not None:
+            value = self.value_to_graph_object(subplot_key, value,
+                                               _raise=_raise)
+            if isinstance(value, (PlotlyDict, PlotlyList)):
+                value.__dict__['_parent'] = self
+                value.__dict__['_parent_key'] = key
+                return super(PlotlyDict, self).__setitem__(key, value)
+
+        if key not in self._attributes:
+            if _raise:
+                raise exceptions.PlotlyDictKeyError(self, key)
+            return
+
+        if graph_objs_tools.get_role(self, key) == 'object':
+            value = self.value_to_graph_object(key, value, _raise=_raise)
+            if isinstance(value, (PlotlyDict, PlotlyList)):
+                value.__dict__['_parent'] = self
+                value.__dict__['_parent_key'] = key
+            else:
+                return
 
-    def _assign_id_to_src(self, src_name, src_value):
-        if isinstance(src_value, six.string_types):
-            src_id = src_value
+        super(PlotlyDict, self).__setitem__(key, value)
+
+    def __getattr__(self, key):
+        """Python only calls this when key is missing!"""
+        try:
+            return self.__getitem__(key)
+        except KeyError:
+            raise AttributeError(key)
+
+    def __copy__(self):
+
+        # TODO: https://github.com/plotly/python-api/issues/291
+        return GraphObjectFactory.create(self._name, **self)
+
+    def __deepcopy__(self, memodict={}):
+
+        # TODO: https://github.com/plotly/python-api/issues/291
+        return self.__copy__()
+
+    def __missing__(self, key):
+
+        if key in self._attributes:
+            if graph_objs_tools.get_role(self, key) == 'object':
+                value = GraphObjectFactory.create(key)
+                value.__dict__['_parent'] = self
+                value.__dict__['_parent_key'] = key
+                return super(PlotlyDict, self).__setitem__(key, value)
+
+        subplot_key = self._get_subplot_key(key)
+        if subplot_key is not None:
+            value = GraphObjectFactory.create(subplot_key)
+            value.__dict__['_parent'] = self
+            value.__dict__['_parent_key'] = key
+            super(PlotlyDict, self).__setitem__(key, value)
+
+    def _get_subplot_key(self, key):
+
+        # TODO: this can use _isSubplotObj instead and won't require subclass!
+        subplot_key_strings = ('xaxis', 'yaxis', 'zaxis', 'lataxis', 'lonaxis',
```

Comment:
Yep. This needs a revisit. I need to go back in and check the `_isSubplotObj`. These hard-coded strings will go away after that.